File: Algorithms/My/Fibonacci/Fibonacci.swift
//
//  Fibonacci.swift
//  DataStructures
//
//  Created by Sicc on 20/10/2019.
//  Copyright © 2019 chang sic jung. All rights reserved.
//

import Foundation

// 반복문 O(N)
public func fibonacci_for(n: Int) -> Int {
  if n == 0 || n == 1 {
    return n
  }
  
  var result = 0
  var a = 0
  var b = 1
  
  for _ in 2...n {
    result = a + b
    a = b
    b = result
  }
  
  return result
}

// 재귀
// 종료조건 만들어야.
// O(2^N)
public func fibonacci_recur(n: Int) -> Int {
  // 종료조건
  if n == 0 || n == 1 {
    return n
  }
  return fibonacci_recur(n: n - 1) + fibonacci_recur(n: n - 2)
}


// 메모제이션
// 공간복잡도를 희생해서 시간복잡도를 높여보자.
// 공간의 max 는 100
// O(N)
var memo = Array(repeating: -1, count: 100)

public func fibonacci_memo(n: Int) -> Int {
  
  if n == 0 || n == 1 {
    memo[n] = n
  }
  
  if memo[n] != -1 { // memo에 값이 있으면.
    return memo[n]
  } else { // 값이 없으면
    memo[n] = fibonacci_memo(n: n - 1) + fibonacci_memo(n: n - 2)
    return memo[n]
  }
  
}

File: Algorithms/My/Fibonacci/Fibonacci.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Oct 20 23:59:16 2019

@author: changsicjung
"""


# 반복문
def fibonacci_for(n):
    if n == 0 or n == 1:
        return n
    result = 0
    a = 0
    b = 1
    
    for i in range(2, n + 1):
        result = a + b
        a = b
        b = result
    return result

# 재귀호출
def fibonacci_recur(n):
    if n == 0 or n == 1:
        return n
    
    return fibonacci_recur(n - 1) + fibonacci_recur(n - 2)



# 메모제이션을 적용한 재귀. 
memo = [-1] * 100
def fibonacci_memo(n):
    if n == 0 or n == 1:
        memo[n] = n
        
    if memo[n] != -1:
        return memo[n]
    else:
        memo[n] = fibonacci_memo(n - 1) + fibonacci_memo(n - 2)
        return memo[n]